### ISSUE TITLE
[distributed] Using gof3r tool can cause inconsistencies in multipart upload

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio#3176]
```
$ gof3r put -p ~/xubuntu-xenial-mate-20160528-longsleep-pine64-8GB.zip -k obj -b testbucket --endpoint=play.minio.io:9000 --path-style
```

```
$ mc ls -I play/testbucket/
[2016-11-04 10:53:37 PDT] 1.4GiB obj
$ mc rm -I -r play/testbucket/
mc: <ERROR> Removal requires --force option. This operational is *IRREVERSIBLE*. Please review carefully before performing this *DANGEROUS* operation. 
$ mc rm -I -r --force play/testbucket/
Removing ‘play/testbucket/obj’.
$ mc rm -I -r play/testbucket/
mc: <ERROR> Removal requires --force option. This operational is *IRREVERSIBLE*. Please review carefully before performing this *DANGEROUS* operation. 
$ mc ls -I play/testbucket/
[2016-11-04 11:03:00 PDT]     0B obj/
$ mc ls -I play/testbucket/
[2016-11-04 11:03:01 PDT]     0B obj/
$ mc ls -I play/testbucket/obj/
mc: <ERROR> Unable to stat ‘play/testbucket/obj/’. Object key is missing, object key cannot be empty
$ mc ls -I play/testbucket/
[2016-11-04 11:03:06 PDT]     0B obj/
$ mc ls -I play/testbucket/
[2016-11-04 11:03:07 PDT]     0B obj/
$ mc rm -I -r play/testbucket/
mc: <ERROR> Removal requires --force option. This operational is *IRREVERSIBLE*. Please review carefully before performing this *DANGEROUS* operation. 
$ mc rm -I -r --force play/testbucket/
Removing ‘play/testbucket/obj’.
$ mc rm -I -r --force play/testbucket/
Removing ‘play/testbucket/obj’.
$ mc ls -I play/testbucket/
[2016-11-04 11:03:18 PDT]     0B obj/
$ mc ls -I play/testbucket/
```

## Expected Behavior
No races are found. 

## Current Behavior
There are races reproduced like this on a common case which we need to fix.

## Possible Solution
Do not have yet. 

## Steps to Reproduce (for bugs)
Just run the gof3r tool with a large file.  Attaching it here as well. 
[gof3r.gz](https://github.com/minio/minio/files/572316/gof3r.gz)

## Context
There are some important races in the multipart codepath which needs to be addressed. 

## Your Environment

Latest master.

**Comment by GarimaKapoor** [comment]
@donatello please update the timeline for the fix.

**Comment by donatello** [comment]
@GarimaKapoor This one appears to be related to the existing multipart issue being worked on in https://github.com/minio/minio/issues/3149. I think it should be fixed in a couple of days.

**Comment by harshavardhana** [comment]
This is related to #3279 the symptoms are pretty similar i think, we can get more 'mc ls' or like output.

**Comment by harshavardhana** [comment]
This is most probably fixed by #3294 - not able to reproduce this issue anymore. Closing for now, will re-open if seen again.
